### PR TITLE
feat: exclusive logging for important stuff

### PR DESCRIPTION
**Pull request by liorzam**
### Description
* We would like to stop toasing not relevant notifications like Unauthorized error

* We started to log our Axios fetch request

    -   Will intercept http service

    -   Api plugin

    -   Saleforce plugin

*  **Workflow Service**
    -   Every-time our backend couldn't fetch an http route url remotely, the server thrown 500 status code instead of 404
    -   We would like reduce redundant:

        **  Logs for incoming request
        
        **  Min log level for application logger

        **   Stop log none important responses for (401/404 status codes)

### Related issues
https://linear.app/ballerine/issue/BAL-888/remove-redundant-toast-in-backoffice-and-better-logging

### CODE REVIEW

File: apps/backoffice-v2/src/common/utils/fetcher/fetcher.ts [hunk 1]
```diff
@@ -49,8 +49,6 @@ export const fetcher: IFetcher = async ({
         : message;
     }
 
-    console.error(message);
-
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/common/utils/fetcher/fetcher.ts]: We might want to at least make the log optional or provide some other indicator for errors.

File: apps/backoffice-v2/src/domains/workflows/hooks/mutations/useUpdateWorkflowByIdMutation/useUpdateWorkflowByIdMutation.tsx [hunk 2]
```diff
@@ -48,8 +48,9 @@ export const useUpdateWorkflowByIdMutation = ({ workflowId }: { workflowId: stri
     onSuccess: (data, { action }) => {
       toast.success(t(`toast:${action}.success`));
     },
-    onError: (error, { action }, context) => {
-      toast.error(t(`toast:${action}.error`, { errorMessage: error.message }));
+    onError: (error: Error, { action }, context) => {
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/domains/workflows/hooks/mutations/useUpdateWorkflowByIdMutation/useUpdateWorkflowByIdMutation.tsx]: Aslong as a check like `instanceof Error` or a function like `isErrorWithMessage` is not in use error is of type `unknown`.

File: apps/backoffice-v2/src/domains/workflows/hooks/mutations/useUpdateWorkflowByIdMutation/useUpdateWorkflowByIdMutation.tsx [hunk 3]
```diff
@@ -48,8 +48,9 @@ export const useUpdateWorkflowByIdMutation = ({ workflowId }: { workflowId: stri
...
+      const translatedError = t(`toast:${action}.error`, { errorMessage: error.message });
+      toast.error(translatedError);
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/domains/workflows/hooks/mutations/useUpdateWorkflowByIdMutation/useUpdateWorkflowByIdMutation.tsx]: A space between line 52 and 53 for readability would be appreciated. 🙏🏼

File: apps/backoffice-v2/src/lib/react-query/query-client.ts [hunk 4]
```diff
@@ -15,6 +15,25 @@ export const isErrorWithCode = (error: unknown): error is IErrorWithCode => {
...
+
+async function clearAuthenticatedUser(queryClient: QueryClient) {
+  try {
+    await queryClient.cancelQueries();
+
+    queryClient.clear();
+
+    const authenticatedUser = authQueryKeys.authenticatedUser();
+
+    queryClient.setQueryData(authenticatedUser.queryKey, {
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/lib/react-query/query-client.ts:17]: Does this matter after `queryClient.cancelQueries`?

File: apps/backoffice-v2/src/lib/react-query/query-client.ts [hunk 5]
```diff
@@ -15,6 +15,25 @@ export const isErrorWithCode = (error: unknown): error is IErrorWithCode => {
   return isObject(error) && 'code' in error && typeof error.code === 'number';
 };
 
+export const isErrorWithMessage = (error: unknown): error is IErrorWithMessage => {
+  return isObject(error) && 'message' in error && typeof error.message === 'string';
+};
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/lib/react-query/query-client.ts]: Why not use the version from `@ballerine/common`?

File: apps/backoffice-v2/src/lib/react-query/query-client.ts [hunk 6]
```diff
@@ -27,28 +46,32 @@ export const queryClient = new QueryClient({
...
-
         return;
       }
 
-      if (!isErrorWithMessage(error) || error.message === 'undefined' || error.message === 'null')
+      if (!isErrorWithMessage(error) || error.message === 'undefined' || error.message === 'null') {
         return;
+      }
+
+      if (isErrorWithCode(error)) {
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/lib/react-query/query-client.ts]: Is there a likelihood that an error would have a code but no message or vice versa and we'd still like to render a toast?

File: services/workflows-service/src/common/filters/AllExceptions.filter.ts [hunk 7]
```diff
@@ -10,19 +14,86 @@ export class AllExceptionsFilter extends BaseExceptionFilter {
...
+      status,
+      error: inspect(errorRes),
+      message: error.message,
+      responseTime: Date.now() - request.startTime,
     });
   }
+
+  private _isObject(obj: unknown): obj is object {
+    return obj !== null && typeof obj === 'object';
+  }
```
Comment (Omri-Levy) [inline services/workflows-service/src/common/filters/AllExceptions.filter.ts]: We already have many versions of this function. Please use an existing option from either `@ballerine/common` or `lodash`.

File: services/workflows-service/src/common/filters/session-exception.filter.ts [hunk 8]
```diff
@@ -20,8 +20,8 @@ export class SessionExpiredExceptionFilter extends BaseExceptionFilter {
       const asyncLogout = util.promisify(req.logout.bind(req));
       await asyncLogout();
 
-      response.clearCookie('session');
-      response.clearCookie('session.sig');
+      // response.clearCookie('session');
+      // response.clearCookie('session.sig');
```
Comment (Omri-Levy) [inline services/workflows-service/src/common/filters/session-exception.filter.ts]: Maybe we should add a comment as for why this is commented out?

File: apps/backoffice-v2/src/lib/react-query/query-client.ts [hunk 9]
```diff
@@ -27,28 +46,32 @@ export const queryClient = new QueryClient({
...
+          void clearAuthenticatedUser(queryClient);
+        }
 
-      toast.error(error.message, {
-        id: error.message,
-      });
+        // Dont toast for no important errors
+        if (
+          statusCode === undefined ||
+          (statusCode >= 500 && statusCode !== 401 && statusCode !== 404)
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/lib/react-query/query-client.ts]: `statusCode` cannot be both equal or greater 500 and in the 400s. Additionally, with this change we're no longer displaying validation errors nor do we have an alternative.

File: services/workflows-service/src/storage/storage.controller.internal.ts [hunk 10]
```diff
@@ -137,15 +154,15 @@ export class StorageControllerInternal {
...
+    const downloadedFilePath = await new HttpFileService({
+      client: this.httpService,
+      logger: this.logger,
+    }).download(persistedFile.uri, localeFilePath);
 
     return downloadedFilePath;
   }
 
-  __isImageUrl(persistedFile: File) {
+  _isUriScheme(persistedFile: File) {
```
Comment (Omri-Levy) [inline services/workflows-service/src/storage/storage.controller.internal.ts]: Also, since its `z.string().url()` why was this change made?
```suggestion
  _isUri(persistedFile: File) {
```

File: services/workflows-service/src/health/consts.ts [hunk 11]
```diff
@@ -0,0 +1,6 @@
+export const CONTROLLER_NAME = '_health';
+
+export const ROUTES = {
+  LIVE: 'live',
+  READY: 'ready',
+};
```
Comment (Omri-Levy) [inline services/workflows-service/src/health/consts.ts]: Note this is the syntax for enums as objects.
```suggestion
} as const;
```

File: apps/backoffice-v2/src/common/utils/handle-promise/handle-promise.ts [hunk 12]
```diff
@@ -1,6 +1,6 @@
 export const handlePromise = async <TData>(
   promise: Promise<TData>,
-): Promise<[TData, undefined] | [undefined, Error]> => {
+): Promise<[Response<TData>, undefined] | [undefined, Error]> => {
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/common/utils/handle-promise/handle-promise.ts]: `handlePromise` handles any promise and not only fetch/http related promises. If a need came up then its still possible to pass a type as a type argument.

File: services/workflows-service/src/common/interceptors/log-request.interceptor.ts [hunk 13]
```diff
@@ -1,23 +1,45 @@
...
+  HttpStatus,
+  Injectable,
+  NestInterceptor,
+} from '@nestjs/common';
+import { isAxiosError } from 'axios';
+import { catchError, tap, throwError } from 'rxjs';
 
 @Injectable()
-export class LogRequestInterceptor implements NestInterceptor {
+export class AxiosRequestErrorInterceptor implements NestInterceptor {
```
Comment (liorzam) [inline services/workflows-service/src/common/interceptors/log-request.interceptor.ts]: rename the file name

File: services/workflows-service/src/common/http-service/http-config.service.ts [hunk 14]
```diff
@@ -0,0 +1,55 @@
+import { HttpModule, HttpModuleOptions, HttpModuleOptionsFactory } from '@nestjs/axios';
+import { HttpStatus, Injectable } from '@nestjs/common';
+import { ConfigModule, ConfigService } from '@nestjs/config';
+import { AppLoggerService } from '../app-logger/app-logger.service';
+
+export const RETRY_DELAY_IN_MS = 1500; // extract to config, delay retry
+
+const _httpStatusFromAxiosMap = {
```
Comment (liorzam) [inline services/workflows-service/src/common/http-service/http-config.service.ts]: use this mapping in logger interceptor

File: services/workflows-service/src/common/middlewares/request-id.middleware.ts [hunk 15]
```diff
@@ -1,38 +1,76 @@
...
+    };
+
+    const abortFn = () => {
+      cleanup();
+      this.logger.warn('Request aborted by the client');
+    };
+
+    const errorFn = (error: Error) => {
+      cleanup();
+      this.logger.log(`Request pipeline error`, { error });
```
Comment (liorzam) [inline services/workflows-service/src/common/middlewares/request-id.middleware.ts:57]: logger.error

**Review by alonp99 — APPROVED** [review]

File: apps/backoffice-v2/src/lib/react-query/query-client.ts (hunk 4 above)
Comment (liorzam) [reply apps/backoffice-v2/src/lib/react-query/query-client.ts:17]: Not sure, but i supposed it should be here because some component bind to this query client state ( so I wonder when use cancel if related pages will get rendered)

will check it

File: apps/backoffice-v2/src/lib/react-query/query-client.ts (hunk 6 above)
Comment (liorzam) [reply apps/backoffice-v2/src/lib/react-query/query-client.ts]: you're right it shouldnt happened, but we are not charge of all the errors that's come up from the query client,
so in case some other issue which is not related to HTTP operation will raise we still would like to show the error message

File: services/workflows-service/src/common/filters/AllExceptions.filter.ts (hunk 7 above)
Comment (liorzam) [reply services/workflows-service/src/common/filters/AllExceptions.filter.ts]: yeah you can see no one is using this method, removed. 
thanks

Comment (liorzam) [reply services/workflows-service/src/common/filters/AllExceptions.filter.ts]: i used import `{ isObject } from '@ballerine/common';`

File: apps/backoffice-v2/src/lib/react-query/query-client.ts (hunk 9 above)
Comment (liorzam) [reply apps/backoffice-v2/src/lib/react-query/query-client.ts]: yeah you are right it wont cover any 4XX status
Do you think we should show 400 error code, i wonder if its not the developer fault instead of`Bad Request` toast which tell nothing to the customer

File: apps/backoffice-v2/src/lib/react-query/query-client.ts (hunk 6 above)
Comment (liorzam) [reply apps/backoffice-v2/src/lib/react-query/query-client.ts]: what you would expect us to do?
in that case we have an issue that we cant provide a readable message for our customers

Comment (liorzam) [reply apps/backoffice-v2/src/lib/react-query/query-client.ts]: the logic is:
if we dont have a message we wont show any toast,
in case we have a message we still will check the status code to make sure its a code we care to tell the user about.

if both not permit its a problem because something wrong happened and we didnt mentioned it, a refresh can might help in this scenario ?

File: apps/backoffice-v2/src/common/utils/fetcher/fetcher.ts [hunk 16]
```diff
@@ -15,13 +15,21 @@ export const fetcher: IFetcher = async ({
...
   isBlob = false,
+}: {
+  url: string;
+  method: string;
+  body: BodyInit | null;
+  headers: HeadersInit;
+  timeout: number;
+  options: Partial<RequestInit>;
+  isBlob: boolean;
 }) => {
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/common/utils/fetcher/fetcher.ts]: `fetcher` receives its types from `IFetcher`. If the types are broken then we need to fix `IFetcher` instead.

File: apps/backoffice-v2/src/lib/react-query/query-client.ts (hunk 4 above)
Comment (liorzam) [reply apps/backoffice-v2/src/lib/react-query/query-client.ts:17]: its seems its works like that all over the application

File: apps/backoffice-v2/src/common/utils/fetcher/fetcher.ts (hunk 16 above)
Comment (liorzam) [reply apps/backoffice-v2/src/common/utils/fetcher/fetcher.ts]: reverted

File: apps/backoffice-v2/src/lib/react-query/query-client.ts [hunk 17]
```diff
@@ -27,28 +33,34 @@ export const queryClient = new QueryClient({
...
-      if (isErrorWithCode(error) && error.code === 401) {
-        const authenticatedUser = authQueryKeys.authenticatedUser();
-
-        void queryClient.cancelQueries();
-        queryClient.setQueryData(authenticatedUser.queryKey, {
-          user: undefined,
-        });
-      }
-
+    onError: (error: unknown) => {
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/lib/react-query/query-client.ts]: `error` is inferred. In general, it's better to prefer inference (for return type as well) unless it's a specific library code where there are consumers for the type or code that requires a specific signature, those being the exception, not the rule.

File: services/workflows-service/tsconfig.json [hunk 18]
```diff
@@ -3,6 +3,7 @@
   "compilerOptions": {
     "emitDecoratorMetadata": true,
     "experimentalDecorators": true,
+    "useUnknownInCatchVariables": false,
```
Comment (Omri-Levy) [inline services/workflows-service/tsconfig.json]: Revert this change, please. 🙏🏼

File: apps/backoffice-v2/src/lib/react-query/query-client.ts [hunk 19]
```diff
@@ -27,28 +33,34 @@ export const queryClient = new QueryClient({
...
+          error.message !== 'null'
+        ) {
+          // Dont toast for no important errors
+          if (![401, 403, 404].includes(status)) {
+            toast.error(error.message, {
+              id: error.message,
+            });
+          }
+        }
+      }
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/lib/react-query/query-client.ts]: This block is getting hard to read. My suggestion is the extreme to the other direction so feel free to say if you see a middle-ground.
```suggestion
if (!isErrorWithCode(error)) return;

const status = error.code;

if (status === 401) {
  void clearAuthenticatedUser(queryClient);
}

if (!isErrorWithMessage(error) || error.message === 'undefined' || error.message === 'null') {
  return;
}

// Dont toast for no important errors
if ([401, 403, 404].includes(status)) return;

toast.error(error.message, {
  id: error.message,
});
```

File: apps/backoffice-v2/src/lib/react-query/query-client.ts [hunk 20]
```diff
@@ -27,24 +33,27 @@ export const queryClient = new QueryClient({
...
+
+      if (!isErrorWithCode(error)) throw error;
+
+      const status = error.code;
+
+      // Dont toast for no important errors
+      if ([401, 403, 404].includes(status)) {
+        if (status === 401) {
+          void clearAuthenticatedUser(queryClient);
+        }
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/lib/react-query/query-client.ts]: There's no reason for this if condition to be nested.

File: apps/backoffice-v2/src/lib/react-query/query-client.ts [hunk 21]
```diff
@@ -27,24 +33,27 @@ export const queryClient = new QueryClient({
...
-      if (isErrorWithCode(error) && error.code === 401) {
-        const authenticatedUser = authQueryKeys.authenticatedUser();
-
-        void queryClient.cancelQueries();
-        queryClient.setQueryData(authenticatedUser.queryKey, {
-          user: undefined,
-        });
-      }
-
+    onError: (error: unknown) => {
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/lib/react-query/query-client.ts]: My previous comment was resolved but the change have not been made. 🙂 
```suggestion
    onError: (error) => {
```

File: services/workflows-service/jest.config.cjs [hunk 22]
```diff
@@ -5,7 +5,6 @@ module.exports = {
   testRegex: '(/__tests__/.*|(\\.|/)(unit|e2e|intg)\\.test)\\.ts$',
   moduleNameMapper: {
     '^@/(.*)$': '<rootDir>/src/$1',
-    axios: 'axios/dist/node/axios.cjs',
```
Comment (Omri-Levy) [inline services/workflows-service/jest.config.cjs:8]: Without this line tests failed due to Axios. Is this no longer the case?

File: services/workflows-service/package.json [hunk 23]
```diff
@@ -15,20 +15,19 @@
...
-    "test": "jest --verbose",
+    "test": "jest",
     "test:unit": "cross-env SKIP_DB_SETUP_TEARDOWN=true jest --testRegex '.*\\.unit\\.test\\.ts$'",
     "test:integration": "jest --testRegex '.*\\.intg\\.test\\.ts$'",
     "test:e2e": "jest --testRegex '.*\\.e2e\\.test\\.ts$'",
     "test:watch": "jest --verbose --watch",
     "seed": "ts-node scripts/seed.ts",
     "seed:ex": "ts-node scripts/seed-ex.ts",
-    "db:migrate-save": "prisma migrate dev",
+    "db:migrate-dev": "prisma migrate dev",
```
Comment (Omri-Levy) [inline services/workflows-service/package.json:25]: Good change! Worth notifying the team it has been made. 🙏🏼

**Comment by liorzam** [comment]
In order to solve the test issues you can follow this link which specify what actually have done in order to fix those tests:

https://stackoverflow.com/questions/76491001/unit-testing-services-that-use-httpmodule/77676262#77676262

File: services/workflows-service/package.json (hunk 23 above)
Comment (liorzam) [reply services/workflows-service/package.json:25]: no one actually use it

Comment (liorzam) [reply services/workflows-service/package.json:25]: it usually used when running db:init
```
"db:init": "npm run db:migrate-dev -- --name 'initial version' && npm run db:migrate-up seed",
```

File: services/workflows-service/src/common/http-service/utils.ts [hunk 24]
```diff
@@ -0,0 +1,48 @@
+import { HttpException, HttpStatus } from '@nestjs/common';
+import { AxiosError, isAxiosError } from 'axios';
+
+const _httpStatusFromAxiosMap = {
```
Comment (Omri-Levy) [inline services/workflows-service/src/common/http-service/utils.ts]: Note the `_` prefix suggests this map is not in use.
```suggestion
const httpStatusFromAxiosMap = {
```

File: services/workflows-service/src/common/http-service/utils.ts [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+  ENOTFOUND: HttpStatus.NOT_FOUND,
+  ECONNABORTED: HttpStatus.INTERNAL_SERVER_ERROR,
+  ERR_CANCELED: HttpStatus.INTERNAL_SERVER_ERROR, // connection canceled
+  ERR_NETWORK: HttpStatus.INTERNAL_SERVER_ERROR, // connection problems
+};
```
Comment (Omri-Levy) [inline services/workflows-service/src/common/http-service/utils.ts]: Remember to add `as const` to maps. 🙂 
```suggestion
} as const;
```

File: services/workflows-service/src/providers/file/file-provider/http-file.service.ts [hunk 26]
```diff
@@ -1,45 +1,77 @@
...
   }
 
+  // private _shouldRetry(error: AxiosError) {
+  //   const { response } = error;
+  //   const isTooManyRequests =
+  //     response?.status === HttpStatusCode.TooManyRequests;
+  //   const isNetworkIssue = error.code === 'ENETUNREACH';
+
+  //   return isTooManyRequests || isNetworkIssue;
+  // }
```
Comment (Omri-Levy) [inline services/workflows-service/src/providers/file/file-provider/http-file.service.ts:53]: Should these comments be removed?

File: services/workflows-service/src/test/helpers/nest-app-helper.ts [hunk 27]
```diff
@@ -46,7 +50,9 @@ const aclFilterResponseInterceptor = {
 export const fetchServiceFromModule = async <T>(
   service: Type<T>,
   dependencies: Provider[] = [],
-  modules: Type<any>[] = [],
+  modules:
+    | Array<Type<any> | DynamicModule | Promise<DynamicModule> | ForwardReference>
+    | Type<unknown>[] = [],
```
Comment (Omri-Levy) [inline services/workflows-service/src/test/helpers/nest-app-helper.ts:55]: Appreciated! 🙌🏼

File: services/workflows-service/package.json (hunk 23 above)
Comment (Omri-Levy) [reply services/workflows-service/package.json:25]: It is used when making a change to the Prisma schema.

File: services/workflows-service/src/common/http-service/http-config.service.ts [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+export class HttpConfigService implements HttpModuleOptionsFactory {
+  constructor(
+    private readonly configService: ConfigService,
+    private readonly logger: AppLoggerService,
+  ) {}
+
+  createHttpOptions(): HttpModuleOptions {
+    return {
+      timeout: this.configService.get('HTTP_TIMEOUT_IN_MS', 5000),
+      maxRedirects: this.configService.get('HTTP_MAX_REDIRECTS', 10),
```
Comment (Omri-Levy) [inline services/workflows-service/src/common/http-service/http-config.service.ts:18]: Does it not make sense to use `env` from @t3-env here?

File: services/workflows-service/src/providers/file/file-provider/http-file.service.ts (hunk 26 above)
Comment (liorzam) [reply services/workflows-service/src/providers/file/file-provider/http-file.service.ts:53]: because this PR getting to big so i decided to add it as technical debt, this comment is still important

File: services/workflows-service/src/common/http-service/http-config.service.ts (hunk 28 above)
Comment (liorzam) [reply services/workflows-service/src/common/http-service/http-config.service.ts:18]: oh no this is the wrong way,
we should alway alway use ConfigService instead t3-env file...

As you can see here i already migrated the [env to our configService](https://github.com/ballerine-io/ballerine/blob/dev/services/workflows-service/src/app.module.ts#L64)
image

File: services/workflows-service/src/common/http-service/utils.ts (hunk 24 above)
Comment (liorzam) [reply services/workflows-service/src/common/http-service/utils.ts]: is it in used here

**Review by Omri-Levy — APPROVED** [review]
LGTM
